### PR TITLE
added support for detecting multiple namespaces in a layer

### LINKED ISSUE #150: detect multiply namespaces in a layer
I met this issue when I work on https://github.com/coreos/clair/issues/40.

The current concept of 'namespace' is inherited from 'operating system', so the `debian/redhat/ubuntu` data source which will never coexist with each other in a layer. But if we count npm/python in, we will have multiple namespaces in a layer. 
 I think we need to modify currently scanning process a bit like this:
1. detect multiple namespaces
     `namespace.go:DetectNamespace()` should return a list of `database.Namespace`
2. detect multiple data if we find multiple namespaces
3. scan the layer by multiple namespace database
   layer struct { `Namespace     *Namespace` } should be a list of `database.Namespace`

We can explain 'namespace' in this way: a self-consistent package system. Although `Namespace - a context around features and vulnerabilities` is a better explanation.

### PR DESCRIPTION
**Pull request by KeyboardNerd**
created table layer_namespace to store the many to many unique mapping of layers and namespaces
changed v1 api to provide a list of namespaces for each layer
changed namespace detector to use all registered detectors to detect namespaces
updated tests for multiple namespaces

Fixes #150

### CODE REVIEW

File: database/models.go [hunk 1]
```diff
@@ -30,9 +30,14 @@ type Layer struct {
...
-	Namespace     *Namespace
-	Features      []FeatureVersion
+
+
+	Parent *Layer
+
+	Namespaces []Namespace
+
+	Features []FeatureVersion
+
```
Comment (jzelinskie) [inline database/models.go]: Any reason why you spaced these out like this? Did this pass gofmt?

File: database/pgsql/layer.go [hunk 2]
```diff
@@ -313,14 +326,37 @@ func (pgSQL *pgSQL) InsertLayer(layer database.Layer) error {
...
 		}
 
+		// replace the old namespace in the database
+		_, err := tx.Exec(removeLayerNamespace, layer.ID)
+		if err != nil {
+			tx.Rollback()
+			return handleError("removeLayerNamespace", err)
+		}
+		for nsid := range namespaceIDs {
+			_, err := tx.Exec(insertLayerNamespace, layer.ID, nsid)
```
Comment (jzelinskie) [inline database/pgsql/layer.go]: This could all be done with one insert rather one query per namespace

File: database/pgsql/queries.go [hunk 3]
```diff
@@ -77,12 +77,17 @@ const (
...
-			LEFT JOIN Namespace n ON l.namespace_id = n.id
+			SELECT l.id, l.name, l.engineversion, p.id, p.name
+			FROM Layer l
+				LEFT JOIN Layer p ON l.parent_id = p.id
 		WHERE l.name = $1;`
 
+	searchLayerNamespace = `SELECT n.id, n.name, n.version_format 
+							FROM Namespace n
+								JOIN Layer_Namespace lns ON lns.namespace_id = n.id 
+							WHERE lns.layer_id = $1
```
Comment (jzelinskie) [inline database/pgsql/queries.go]: indentation is a little funky here

**Review by jzelinskie — CHANGES_REQUESTED** [review]
awesome work! mostly just nitpicks

File: database/models.go (hunk 1 above)
Comment (KeyboardNerd) [reply database/models.go]: ah, the broken gofmt in emacs caused this

File: database/pgsql/migrations/00007_add_multiplens.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrations
+
+import "github.com/remind101/migrate"
+
+func init() {
+	RegisterMigration(migrate.Migration{
+		ID: 7,
```
Comment (jzelinskie) [inline database/pgsql/migrations/00007_add_multiplens.go]: This needs to change to 8.

File: database/pgsql/layer.go [hunk 5]
```diff
@@ -329,6 +348,21 @@ func (pgSQL *pgSQL) InsertLayer(layer database.Layer) error {
...
+	stmt, err := tx.Prepare(insertLayerNamespace)
+	for nsid := range namespaceIDs {
+		_, err := stmt.Exec(layer.ID, nsid)
+		if err != nil {
+			tx.Rollback()
+			return handleError("insertLayerNamespace", err)
+		}
+	}
+
+	err = stmt.Close()
```
Comment (jzelinskie) [inline database/pgsql/layer.go]: If anything fails while iterating over the namespaceIDs, the statement never gets closed. I think it's best to structure it like this instead:
```go
stmt, err := tx.Prepare(insertLayerNamespace)
defer func() {
  err = stmt.Close()
  if err != nil {
    tx.Rollback()
    log.WithFields(log.Fields{"error": err}).Error("failed to close prepared statement")
}()

for nsid := range namespaceIDs {
  ...
}
```

File: database/pgsql/queries.go [hunk 6]
```diff
@@ -77,12 +77,18 @@ const (
...
+			FROM Layer l
+				LEFT JOIN Layer p ON l.parent_id = p.id
 		WHERE l.name = $1;`
 
+	searchLayerNamespace = `
+			SELECT n.id, n.name, n.version_format 
+			FROM Namespace n
+				JOIN Layer_Namespace lns ON lns.namespace_id = n.id 
+			WHERE lns.layer_id = $1
+							`
```
Comment (jzelinskie) [inline database/pgsql/queries.go]: let's move this ` to right after the $1

File: database/pgsql/testdata/data.sql [hunk 7]
```diff
@@ -28,12 +28,19 @@ INSERT INTO featureversion (id, feature_id, version) VALUES
...
-  (5, 'layer-3b', 1, 3, 2);
+INSERT INTO layer (id, name, engineversion, parent_id) VALUES
+  (1, 'layer-0', 1, NULL),
+  (2, 'layer-1', 1, 1),
+  (3, 'layer-2', 1, 2),
+  (4, 'layer-3a', 1, 3),
+  (5, 'layer-3b', 1, 3);
+
+INSERT INTO layer_namespace (id, layer_id, namespace_id) VALUES
+(1, 2, 1),
```
Comment (jzelinskie) [inline database/pgsql/testdata/data.sql]: indent

File: ext/featurens/driver_test.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+func TestMultipleNamespaceDetector(t *testing.T) {
+	testData := []MultipleNamespaceTestData{
+		{
+			ExpectedNamespaces: []database.Namespace{
+				database.Namespace{Name: "debian:8", VersionFormat: "dpkg"},
+				database.Namespace{Name: "alpine:v3.3", VersionFormat: "dpkg"},
+			},
+			Files: tarutil.FilesMap{
+				"etc/os-release": []byte(
+					`PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
```
Comment (jzelinskie) [inline ext/featurens/driver_test.go]: put the ` on the line above and start PRETTYNAME on a newline right above NAME

**Review by jzelinskie — CHANGES_REQUESTED** [review]

File: database/pgsql/layer.go [hunk 9]
```diff
@@ -329,6 +348,30 @@ func (pgSQL *pgSQL) InsertLayer(layer database.Layer) error {
...
+
+	defer func() {
+		err = stmt.Close()
+		if err != nil {
+			tx.Rollback()
+			log.WithError(err).Error("failed to close prepared statement")
+		}
+	}()
+
+	if err != nil {
```
Comment (jzelinskie) [inline database/pgsql/layer.go]: move this above the defer or else the defer could possibly cause a nil ptr deference

**Review by jzelinskie — CHANGES_REQUESTED** [review]

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**